Remove redundant constraints take 3

vladciobanu opened the pull request:
Fixes #705 

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`
- [ ] Style conformance: `stylish-haskell`

---

vladciobanu commented:
It might be worth it to move the code I added in `Step.Step` to a different module, but I would like a review now that focuses on the changes to the tests.

ttuegel commented on kore/src/Kore/Step/Step.hs:
```diff
@@ -224,8 +229,26 @@ unifyRule
+        Substitution.normalizeExcept nonRedundantPredicate
+
+    -- | ⌈L ∧ φ⌉ ∧ P → Pre
+    makePredicate unificationSolution requires =
```
The comment helps, but the name `makePredicate` is a little vague: what kind of predicate does this make?

ttuegel commented on kore/src/Kore/Step/Step.hs:
```diff
@@ -224,8 +229,26 @@ unifyRule
+    unification' <- removeRedundancy unification requires'
     return (rule' `Conditional.withCondition` unification')
+  where
+    removeRedundancy unificationSolution requires = do
```
The name `removeRedundancy` is unclear about what is redundant. Maybe, `removeRedundantRequires`?

virgil-serbanuta replied:
Actually, wouldn't `addRequiresUnlessRedundant` be a better name? This does not remove anything...

ttuegel commented on kore/test/Test/Kore/OnePath/Step.hs:
```diff
@@ -312,19 +312,9 @@ test_onePathStrategy =
                     $ Mock.f Mock.b
             ]
         assertEqualWithExplanation ""
-            [ Stuck Conditional
```
This change does not seem reasonable to me. We cannot unconditionally prove the claim `constr10(b) => a` given only the axiom `constr10(b) => a requires f(b) = c`.

ttuegel commented on kore/test/Test/Kore/Step/Axiom/EvaluationStrategy.hs:
```diff
@@ -80,18 +82,11 @@ test_definitionEvaluation =
                         OrPattern.fromPattern Conditional
                             { term = Mock.g Mock.a
-                            , predicate = requirement
+                            , predicate = makeTruePredicate
```
Where in the pre-condition does it imply that `f(a) = g(b)`?

ttuegel commented on kore/test/Test/Kore/Step/Function/Integration.hs:
```diff
@@ -1028,5 +1032,42 @@ evaluate
+        }
+  where
+    tools =
+        Mock.makeMetadataTools
```
IIRC, `Mock.makeMetadataTools` does some things that are sketchy, and I intentionally avoided it for this module. Why is this necessary?

ttuegel commented on kore/test/Test/Kore/Step/Axiom/UserDefined.hs:
```diff
@@ -107,14 +107,14 @@ test_userDefinedFunction =
+                    { results = OrPattern.fromPatterns
                         [ Conditional
-                            { term = Mock.functionalConstr10 (mkVar Mock.x)
+                            { term = Mock.functionalConstr11 (mkVar Mock.x)
```
The axiom below cannot apply because it `requires False`.

ttuegel commented on kore/test/Test/Kore/Step/Step.hs:
```diff
@@ -546,12 +546,6 @@ test_applyRewriteRule_ =
         actual <- applyRewriteRule_ initial axiomEnsuresBottom
         assertEqualWithExplanation "" expect actual
 
-    , testCase "rule a => b requires \\bottom" $ do
```
Why was this removed?

ttuegel commented on kore/test/Test/Kore/Step/Axiom/EvaluationStrategy.hs:
```diff
@@ -272,48 +262,6 @@ test_firstFullEvaluation =
                 )
                 (Mock.functionalConstr10 (mkVar Mock.x))
         assertEqualWithExplanation "" expect actual
-    , testCase "Error when not fully rewriting" $ do
```
These errors should not have gone away.

ttuegel commented on kore/test/Test/Kore/Step/Axiom/EvaluationStrategy.hs:
```diff
@@ -352,18 +300,11 @@ test_simplifierWithFallback =
                     { results = OrPattern.fromPatterns
                         [ Conditional
                             { term = Mock.g Mock.a
-                            , predicate = requirement
```
I don't think these tests should have changed, either.

virgil-serbanuta commented on kore/src/Kore/Step/Step.hs:
```diff
@@ -224,8 +229,26 @@ unifyRule
+        isPreRedundant <- evaluate predicate
+        let nonRedundantPredicate =
+                case isPreRedundant of
+                    Just True -> unificationSolution
```
So, I'm a bit confused by this. I think we only implemented the "Just False" case, see https://github.com/kframework/kore/blob/master/kore/src/Kore/Step/SMT/Evaluator.hs#L143
so I would expect this to be a no-op. On the other hand, your change seems to make a difference for the tests below.

So, I'm not sure what comment to make here, but this does not seem right.

vladciobanu commented:
@ttuegel Yep, you are exactly right. I somehow convinced myself that those were right, but they were obviously not. The problem was that I tried to be cute with `Function.&` and accidentally reversed the implication of the predicate, thus making it `Pre → ⌈L ∧ φ⌉ ∧ P` which turns out to be true pretty often.

Additionally, @virgil-serbanuta was correct: we only have the `Just False` case for `Unsat`, so I needed to negate the predicate anyway.

I rebased/force push-ed the new version which looks a lot better now, I think (no change needed to other tests, added a test that is failing on `master` but working here; can you think of other tests I should add?).

@virgil-serbanuta also had an interesting idea I am inclined to take on: what if we change `evaluate` (this tricked me since it returns `Maybe Bool` but it only really returns `Nothing` or `Just False`) to be more precise? For example, we could add some argument that specifies whether we want to prove something's always true or always false and it would negate the predicate itself if needed...?
